lieer/labels_translation.py

```diff
@@ -0,0 +1,332 @@
+
+Usage:
+
+After creating an instance of LabelTranslator, it is necessary to call
+the instance's `load_user_translation` method with an argument that is
+a path of a file containing a translation map and filter patterns
+definitions.
+
+This file should be a JSON file having the following structure:
+
+{
+    "labels_map": {
+        "<remote label>" : "<local tag>",
+        "<remote label>" : "<local tag>",
+        .
+        .
+        .
+    },
+    "ignore_patterns": [
+        "<pattern>",
+        "<pattern>",
+        .
+        .
+        .
+    ],
+    "label_sep": "<label separator>"
+}
+
+Where:
+
+* "label_map" is a mapping between remote labels and local tags.
+
+* "ignore_patterns" is a list of regular expressions; local tags
+  matching any of these patterns will not be synced to Gmail.
+
+* "label_sep" is a string that will replace, in local tags, the slash
+  used by Gmail as a label/sublabel separator.
+
+For example. a custom translation file might look like this:
+
+{
+    "labels_map": {
+        "TRASH"     : "deleted",
+        "my-remote": "my-local"
+    },
+    "ignore_patterns": [
+        "new",
+        "par.*"
+    ],
+    "label_sep": "::"
+}
+
+Using this file will have the following effects:
+
+1. Gmail's 'TRASH' label will be translated to the local tag
+   'deleted' and vice versa;
+
+2. Gmail's label 'my-remote' will be translated to the local
+   tag 'my-local' and vice versa;
+
+3. The local tag 'new' will not be synced to Gmail;
+
+4. Any local tag matching the regular expression 'par.*' (i.e., any
+   tag that starts with 'par') will not be synced to Gmail;
+
+5. The '/' character that separates label/sublabel in Gmail will be
+   replaced, in local tags, with the string '::'. For example, a
+   Gmail 'foo/bar' label will be translated to the local tag 'foo::bar'.
+"""
+
+import os
+import json
+import re
+import notmuch
+
+class LabelTranslator:
+    """Translation between Gmail labels and Notmuch tags."""
+    
+    # the default separator between parent/child labels in Gmail
+    GMAIL_LABEL_SEP = '/'
+
+    # the default tranlation of label names between the local
+    # (notmuch) system and the remote (Gmail) system. This translation
+    # map can be overridden by a user's custom map.
+    DEFAULT_REMOTE_TO_LOCAL_MAP = {
+        'INBOX'     : 'inbox',
+        'SPAM'      : 'spam',
+        'TRASH'     : 'trash',
+        'UNREAD'    : 'unread',
+        'STARRED'   : 'flagged',
+        'IMPORTANT' : 'important',
+        'SENT'      : 'sent',
+        'DRAFT'     : 'draft',
+        'CHAT'      : 'chat'
+    }
+
+    # set of Gmail's system labels that should not be created by gmi.
+    # the labels in this set should be written in all lower case letters.
+    SYSTEM_IGNORED_LABELS = {
+        'attachment',
+        'encrypted',
+        'signed',
+        # 'new',
+        'passed',
+        'replied',
+        'muted',
+        'mute',
+        'todo',
+        'trash',
+        'bin',
```

Comment:
`bin` needs to be added to the README caveats section.

Reply:
I agree. (and `archive` too)